shiny/_app.py

```diff
@@ -438,7 +438,9 @@ def _render_page(self, ui: Tag | TagList, lib_prefix: str) -> RenderedHTML:
         ui_res = copy.copy(ui)
         # Make sure requirejs, jQuery, and Shiny come before any other dependencies.
         # (see require_deps() for a comment about why we even include it)
-        ui_res.insert(0, [require_deps(), jquery_deps(), *shiny_deps()])
+        ui_res.insert(
+            0, [require_deps(), jquery_deps(), *shiny_deps(include_css=False)]
```

Comment:
This is problematic if the UI isn't coming from a page function, for example:

```python
from shiny import App, ui

app_ui = ui.div()

def server(input):
    ui.notification_show("Foo")

app = App(app_ui, server)
```

Seems it'd be better to only add `shiny_deps()` if shiny isn't already in `ui_res.get_dependencies()`?

Reply:
The comment just above this line implies that its important the `shiny_deps()` are added here so that they're first on the page.

So we can't skip `shiny_deps()` if it's in `.get_dependencies()` because the ordering is important. If we're going to inspect the `ui` dependencies, we'd have to look through them to find evidence of the page-level dependency having been added, which feels like a leaking abstraction.

Instead, I think it might be appropriate for us to split the shiny dependency into `shiny-js` and `shiny-css`. I don't usually like this approach, but I think it's appropriate here. We'd have attach `shiny-js` early here and `shiny-css` to any component that uses those styles and we'd also need to suppress those component CSS deps from the page-level deps.

*sigh*. On second thought, maybe it'd be easier to check for the `bootstrap` dependency as the "went through a `ui.page_*()` signal.

Reply:
Here's the other wrinkle: Shiny's CSS makes some pretty strong "Bootstrap is available" assumptions.

```python
from shiny import App, ui

app_ui = ui.div()
# app_ui = ui.page_bootstrap(ui.div())

def server(input):
    ui.notification_show("Foo", duration=99, type="message")
    ui.notification_show("Foo", duration=99, type="warning")

app = App(app_ui, server)
```

| Scenario | Output |
|--------|--------|
| With Bootstrap | ![image](https://github.com/posit-dev/py-shiny/assets/5420529/a2624849-1294-4ab9-9c9e-142dfbc643f6) |
| No Bootstrap + `shiny.css` | ![image](https://github.com/posit-dev/py-shiny/assets/5420529/e509d548-8b5c-4853-9a06-68a2f645571c) |
| No Bootstrap or `shiny.css` | ![image](https://github.com/posit-dev/py-shiny/assets/5420529/ee72faf4-680e-4f88-8355-77e78300bd08) |

Reply:
Anyway, I'm going to go with excluding the `shiny.css` if `bootstrap` is included in the `ui` dependencies.

Reply:
That said we still should probably audit components that use styles from `shiny.min.css`

Reply:
Done in 6a6d5089adb1d655882a3f31eabdb4a7566a1014

Reply:
The same changes in https://github.com/posit-dev/py-shiny/commit/6a6d5089adb1d655882a3f31eabdb4a7566a1014https://github.com/posit-dev/py-shiny/commit/6a6d5089adb1d655882a3f31eabdb4a7566a1014 should also be applied to the method below it?

Reply:
> The same changes in [6a6d508](https://github.com/posit-dev/py-shiny/commit/6a6d5089adb1d655882a3f31eabdb4a7566a1014) should also be applied to the method below it?

Actually, I think we should just include the shiny CSS in that case. dbdb8bdf3e21a72d7cbe221b093304b7a92c3bc2